### ISSUE TITLE
Add generic base template file for object detail views

### ISSUE THREAD

**Issue by glennmatthews** [nautobot/nautobot#479]
### Proposed Changes

Create another page template under `nautobot/core/templates/generic/`; specifically this one would be for "object detail view" pages.

### Justification

Pretty much every detail page template has a slight variant of the below boilerplate - here are three examples:

```jinja2
{% block header %}
    <div class="row noprint">
        <div class="col-sm-8 col-md-9">
            <ol class="breadcrumb">
                <li><a href="{% url 'extras:exporttemplate_list' %}">Export Templates</a></li>
                {% if object.name %}
                    <li><a href="{% url 'extras:exporttemplate' pk=object.pk %}">{{ object.name }}</a></li>
                {% endif %}
            </ol>
        </div>
    </div>

    <div class="pull-right noprint">
        {% plugin_buttons object %}
        {% if perms.extras.add_exporttemplate %}
            {% clone_button object %}
        {% endif %}
        {% if perms.extras.change_exporttemplate %}
            {% edit_button object %}
        {% endif %}
        {% if perms.extras.delete_exporttemplate %}
            {% delete_button object %}
        {% endif %}
    </div>

    <h1>{% block title %}{{ object }}{% endblock %}</h1>
    {% include 'inc/created_updated.html' %}
    <div class="pull-right noprint">
        {% custom_links object %}
    </div>

    <ul class="nav nav-tabs">
        <li role="presentation"{% if not active_tab %} class="active"{% endif %}>
            <a href="{{ object.get_absolute_url }}">Export Templates</a>
        </li>
        {% if perms.extras.view_objectchange %}
            <li role="presentation"{% if active_tab == 'changelog' %} class="active"{% endif %}>
                <a href="{% url 'extras:exporttemplate_changelog' pk=object.pk %}">Change Log</a>
            </li>
        {% endif %}
    </ul>

{% endblock %}
```

```jinja2
{% block header %}
    <div class="row noprint">
        <div class="col-sm-8 col-md-9">
            <ol class="breadcrumb">
                <li><a href="{% url 'dcim:region_list' %}">Regions</a></li>
                <li>{{ object }}</li>
            </ol>
        </div>
        <div class="col-sm-4 col-md-3">
            <form action="{% url 'dcim:region_list' %}" method="get">
                <div class="input-group">
                    <input type="text" name="q" class="form-control" />
                    <span class="input-group-btn">
                        <button type="submit" class="btn btn-primary">
                            <span class="mdi mdi-magnify" aria-hidden="true"></span>
                        </button>
                    </span>
                </div>
            </form>
        </div>
    </div>
    <div class="pull-right noprint">
        {% plugin_buttons object %}
        {% if perms.dcim.add_region %}
            {% clone_button object %}
        {% endif %}
        {% if perms.dcim.change_region %}
            {% edit_button object use_pk=1 %}
        {% endif %}
        {% if perms.dcim.delete_region %}
            {% delete_button object use_pk=1 %}
        {% endif %}
    </div>
    <h1>{{ object }}</h1>
    {% include 'inc/created_updated.html' %}
    <div class="pull-right noprint">
        {% custom_links object %}
    </div>
    <ul class="nav nav-tabs">
        <li role="presentation"{% if not active_tab %} class="active"{% endif %}>
            <a href="{{ object.get_absolute_url }}">Region</a>
        </li>
        {% if perms.extras.view_objectchange %}
            <li role="presentation"{% if active_tab == 'changelog' %} class="active"{% endif %}>
                <a href="{% url 'dcim:region_changelog' pk=object.pk %}">Change Log</a>
            </li>
        {% endif %}
    </ul>
{% endblock %}
```

```jinja2
{% block header %}
    <div class="row noprint">
        <div class="col-sm-8 col-md-9">
            <ol class="breadcrumb">
                <li><a href="{% url 'ipam:role_list' %}">Roles</a></li>
                <li>{{ object }}</li>
            </ol>
        </div>
        <div class="col-sm-4 col-md-3">
            <form action="{% url 'ipam:role_list' %}" method="get">
                <div class="input-group">
                    <input type="text" name="q" class="form-control" />
                    <span class="input-group-btn">
                        <button type="submit" class="btn btn-primary">
                            <span class="mdi mdi-magnify" aria-hidden="true"></span>
                        </button>
                    </span>
                </div>
            </form>
        </div>
    </div>
    <div class="pull-right noprint">
        {% plugin_buttons object %}
        {% if perms.ipam.add_role %}
            {% clone_button object %}
        {% endif %}
        {% if perms.ipam.change_role %}
            {% edit_button object %}
        {% endif %}
        {% if perms.ipam.delete_role %}
            {% delete_button object %}
        {% endif %}
    </div>
    <h1>{{ object }}</h1>
    {% include 'inc/created_updated.html' %}
    <div class="pull-right noprint">
        {% custom_links object %}
    </div>
    <ul class="nav nav-tabs">
        <li role="presentation"{% if not active_tab %} class="active"{% endif %}>
            <a href="{{ object.get_absolute_url }}">Role</a>
        </li>
        {% if perms.extras.view_objectchange %}
            <li role="presentation"{% if active_tab == 'changelog' %} class="active"{% endif %}>
                <a href="{% url 'ipam:role_changelog' slug=object.slug %}">Change Log</a>
            </li>
        {% endif %}
    </ul>
{% endblock %}
```

All that really varies from page to page are the breadcrumbs at the top and the permissions associated with the buttons.

**Comment by jathanism** [comment]
Here are all the detail view templates that need to be refactored to use a generic base:

```
$ find nautobot -name '*.html' | egrep -v "(_|buttons|templatetags|project-static|trace|core|inc|widgets|users)"
nautobot/virtualization/templates/virtualization/virtualmachine.html
nautobot/virtualization/templates/virtualization/vminterface.html
nautobot/virtualization/templates/virtualization/clustertype.html
nautobot/virtualization/templates/virtualization/cluster.html
nautobot/virtualization/templates/virtualization/clustergroup.html
nautobot/circuits/templates/circuits/circuit.html
nautobot/circuits/templates/circuits/provider.html
nautobot/circuits/templates/circuits/circuittype.html
nautobot/tenancy/templates/tenancy/tenantgroup.html
nautobot/tenancy/templates/tenancy/tenant.html
nautobot/dcim/templates/dcim/devicetype.html
nautobot/dcim/templates/dcim/rackrole.html
nautobot/dcim/templates/dcim/devicebay.html
nautobot/dcim/templates/dcim/region.html
nautobot/dcim/templates/dcim/powerfeed.html
nautobot/dcim/templates/dcim/interface.html
nautobot/dcim/templates/dcim/poweroutlet.html
nautobot/dcim/templates/dcim/manufacturer.html
nautobot/dcim/templates/dcim/rackgroup.html
nautobot/dcim/templates/dcim/powerport.html
nautobot/dcim/templates/dcim/platform.html
nautobot/dcim/templates/dcim/rack.html
nautobot/dcim/templates/dcim/frontport.html
nautobot/dcim/templates/dcim/consoleserverport.html
nautobot/dcim/templates/dcim/device.html
nautobot/dcim/templates/dcim/cable.html
nautobot/dcim/templates/dcim/device/rearports.html
nautobot/dcim/templates/dcim/device/poweroutlets.html
nautobot/dcim/templates/dcim/device/status.html
nautobot/dcim/templates/dcim/device/devicebays.html
nautobot/dcim/templates/dcim/device/base.html
nautobot/dcim/templates/dcim/device/inventory.html
nautobot/dcim/templates/dcim/device/interfaces.html
nautobot/dcim/templates/dcim/device/frontports.html
nautobot/dcim/templates/dcim/device/consoleserverports.html
nautobot/dcim/templates/dcim/device/powerports.html
nautobot/dcim/templates/dcim/device/consoleports.html
nautobot/dcim/templates/dcim/device/config.html
nautobot/dcim/templates/dcim/consoleport.html
nautobot/dcim/templates/dcim/rackreservation.html
nautobot/dcim/templates/dcim/virtualchassis.html
nautobot/dcim/templates/dcim/powerpanel.html
nautobot/dcim/templates/dcim/site.html
nautobot/dcim/templates/dcim/inventoryitem.html
nautobot/dcim/templates/dcim/rearport.html
nautobot/dcim/templates/dcim/devicerole.html
nautobot/ipam/templates/ipam/prefix.html
nautobot/ipam/templates/ipam/routetarget.html
nautobot/ipam/templates/ipam/aggregate.html
nautobot/ipam/templates/ipam/vlan.html
nautobot/ipam/templates/ipam/vrf.html
nautobot/ipam/templates/ipam/vlangroup.html
nautobot/ipam/templates/ipam/service.html
nautobot/ipam/templates/ipam/ipaddress.html
nautobot/ipam/templates/ipam/role.html
nautobot/ipam/templates/ipam/rir.html
nautobot/extras/templates/extras/status.html
nautobot/extras/templates/extras/gitrepository.html
nautobot/extras/templates/extras/tag.html
nautobot/extras/templates/extras/job.html
nautobot/extras/templates/extras/configcontext.html
nautobot/extras/templates/extras/jobresult.html
nautobot/extras/templates/extras/configcontextschema/base.html
nautobot/extras/templates/extras/configcontextschema/validation.html
nautobot/extras/templates/extras/graphqlquery.html
nautobot/extras/templates/extras/customlink.html
nautobot/extras/templates/extras/objectchange.html
nautobot/extras/templates/extras/computedfield.html
nautobot/extras/templates/extras/configcontextschema.html
nautobot/extras/templates/extras/webhook.html
nautobot/extras/templates/extras/exporttemplate.html
```
There are **71** of them in total.

**Comment by jathanism** [comment]
Completion checklist:

- [ ] nautobot/virtualization/templates/virtualization/virtualmachine.html
- [ ] nautobot/virtualization/templates/virtualization/vminterface.html
- [ ] nautobot/virtualization/templates/virtualization/clustertype.html
- [ ] nautobot/virtualization/templates/virtualization/cluster.html
- [ ] nautobot/virtualization/templates/virtualization/clustergroup.html
- [ ] nautobot/circuits/templates/circuits/circuit.html
- [ ] nautobot/circuits/templates/circuits/provider.html
- [ ] nautobot/circuits/templates/circuits/circuittype.html
- [ ] nautobot/tenancy/templates/tenancy/tenantgroup.html
- [ ] nautobot/tenancy/templates/tenancy/tenant.html
- [ ] nautobot/dcim/templates/dcim/devicetype.html
- [ ] nautobot/dcim/templates/dcim/rackrole.html
- [ ] nautobot/dcim/templates/dcim/devicebay.html
- [ ] nautobot/dcim/templates/dcim/region.html
- [ ] nautobot/dcim/templates/dcim/powerfeed.html
- [ ] nautobot/dcim/templates/dcim/interface.html
- [ ] nautobot/dcim/templates/dcim/poweroutlet.html
- [ ] nautobot/dcim/templates/dcim/manufacturer.html
- [ ] nautobot/dcim/templates/dcim/rackgroup.html
- [ ] nautobot/dcim/templates/dcim/powerport.html
- [ ] nautobot/dcim/templates/dcim/platform.html
- [ ] nautobot/dcim/templates/dcim/rack.html
- [ ] nautobot/dcim/templates/dcim/frontport.html
- [ ] nautobot/dcim/templates/dcim/consoleserverport.html
- [ ] nautobot/dcim/templates/dcim/device.html
- [ ] nautobot/dcim/templates/dcim/cable.html
- [ ] nautobot/dcim/templates/dcim/device/rearports.html
- [ ] nautobot/dcim/templates/dcim/device/poweroutlets.html
- [ ] nautobot/dcim/templates/dcim/device/status.html
- [ ] nautobot/dcim/templates/dcim/device/devicebays.html
- [ ] nautobot/dcim/templates/dcim/device/base.html
- [ ] nautobot/dcim/templates/dcim/device/inventory.html
- [ ] nautobot/dcim/templates/dcim/device/interfaces.html
- [ ] nautobot/dcim/templates/dcim/device/frontports.html
- [ ] nautobot/dcim/templates/dcim/device/consoleserverports.html
- [ ] nautobot/dcim/templates/dcim/device/powerports.html
- [ ] nautobot/dcim/templates/dcim/device/consoleports.html
- [ ] nautobot/dcim/templates/dcim/device/config.html
- [ ] nautobot/dcim/templates/dcim/consoleport.html
- [ ] nautobot/dcim/templates/dcim/rackreservation.html
- [ ] nautobot/dcim/templates/dcim/virtualchassis.html
- [ ] nautobot/dcim/templates/dcim/powerpanel.html
- [ ] nautobot/dcim/templates/dcim/site.html
- [ ] nautobot/dcim/templates/dcim/inventoryitem.html
- [ ] nautobot/dcim/templates/dcim/rearport.html
- [ ] nautobot/dcim/templates/dcim/devicerole.html
- [ ] nautobot/ipam/templates/ipam/prefix.html
- [ ] nautobot/ipam/templates/ipam/routetarget.html
- [ ] nautobot/ipam/templates/ipam/aggregate.html
- [ ] nautobot/ipam/templates/ipam/vlan.html
- [ ] nautobot/ipam/templates/ipam/vrf.html
- [ ] nautobot/ipam/templates/ipam/vlangroup.html
- [ ] nautobot/ipam/templates/ipam/service.html
- [ ] nautobot/ipam/templates/ipam/ipaddress.html
- [ ] nautobot/ipam/templates/ipam/role.html
- [ ] nautobot/ipam/templates/ipam/rir.html
- [ ] nautobot/extras/templates/extras/status.html
- [ ] nautobot/extras/templates/extras/gitrepository.html
- [ ] nautobot/extras/templates/extras/tag.html
- [ ] nautobot/extras/templates/extras/job.html
- [ ] nautobot/extras/templates/extras/configcontext.html
- [ ] nautobot/extras/templates/extras/jobresult.html
- [ ] nautobot/extras/templates/extras/configcontextschema/base.html
- [ ] nautobot/extras/templates/extras/configcontextschema/validation.html
- [ ] nautobot/extras/templates/extras/graphqlquery.html
- [ ] nautobot/extras/templates/extras/customlink.html
- [ ] nautobot/extras/templates/extras/objectchange.html
- [ ] nautobot/extras/templates/extras/computedfield.html
- [ ] nautobot/extras/templates/extras/configcontextschema.html
- [ ] nautobot/extras/templates/extras/webhook.html
- [ ] nautobot/extras/templates/extras/exporttemplate.html

**Comment by itdependsnetworks** [comment]
This would be great!

**Comment by jathanism** [comment]
> This would be great!

I'm working on it right meow! 😼

**Comment by glennmatthews** [comment]
Addressed by #737, will be present in the 1.2.0 release.
